packages/studio-ui/src/web/components/Layout/TopNav/cloud/Step3-deploy.tsx

```diff
@@ -0,0 +1,152 @@
+import axios from 'axios'
+import { isCDMError, UnreachableCaseError } from 'common/errors'
+import React, { useEffect, useReducer } from 'react'
+import { connect } from 'react-redux'
+import { Timeout, toastFailure } from '~/components/Shared/Utils'
+import { RootReducer } from '~/reducers'
+import { Status } from './deploy'
+import { TrainSessions } from './types'
+
+const BASE_NLU_URL = `${window.STUDIO_API_PATH}/nlu`
+
+const computeIsTrained = (trainSessions: TrainSessions) =>
+  Object.values(trainSessions).reduce((trained, session) => trained && session.status === 'done', true)
+
+type State =
+  | {
+      status: 'train_pending'
+    }
+  | { status: 'training_in_progress' }
+  | { status: 'upload_pending' }
+  | { status: 'upload_completed' }
+  | { status: 'upload_failed' }
+
+type Action =
+  | { type: 'training/started' }
+  | { type: 'training/trainSessionsUpdated'; trainSessions: TrainSessions }
+  | { type: 'upload/ended' }
+  | { type: 'upload/failed' }
+
+function reducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'training/started':
+      return { status: 'training_in_progress' }
+    case 'training/trainSessionsUpdated':
+      const isTrained = computeIsTrained(action.trainSessions)
+      if (isTrained) {
+        return { status: 'upload_pending' }
+      } else {
+        return { status: 'training_in_progress' }
+      }
+    case 'upload/ended':
+      return { status: 'upload_completed' }
+    case 'upload/failed':
+      return { status: 'upload_completed' }
+    default:
+      throw new UnreachableCaseError(action)
+  }
+}
+
+interface OwnProps {
+  pat: string
+  workspaceId: string
+  onCompleted: () => void
+}
+
+type StateProps = ReturnType<typeof mapStateToProps>
+type DispatchProps = typeof mapDispatchToProps
+type Props = DispatchProps & StateProps & OwnProps
+
+function init(props: Props): State {
+  const { trainSessions } = props
+
+  const isTrained = computeIsTrained(trainSessions)
+
+  if (isTrained) {
+    return { status: 'upload_pending' }
+  } else {
+    return { status: 'train_pending' }
+  }
+}
+
+export const Deploy = (props: Props): JSX.Element => {
+  const { pat, trainSessions, workspaceId, onCompleted } = props
+
+  const [state, dispatch] = useReducer(reducer, props, init)
+
+  const { status } = state
+
+  useEffect(() => {
+    const ac = new AbortController()
+    const startTrain = async () => {
+      for (const lang of Object.keys(trainSessions)) {
+        await axios.post(`${BASE_NLU_URL}/train/${lang}`, null, { signal: ac.signal })
+      }
```

Comment:
**question** is there any cases where you have a multilingual chatbot that has no train session for one or more languages ? 

In other words, should we use another source to start training in all languages supported by the chatbot ?
Or should we implement a post to /train which would train all languages at once in the backend and handle that logic in the backend

Reply:
These are all valid questions, but I thought they fell out of scope of this PR.
Personally I would prefer a single `/train` endpoint that would train all languages, but this would have impacts outside the scope of this PR (e.g. the status bar)

Reply:
I'll check if TrainSessions include all lang keys, all the time

Reply:
https://github.com/botpress/studio/blob/df96e6e5be31c690cabff83e1d72b1898f20531d/packages/studio-ui/src/web/components/Layout/StatusBar/TrainingStatus/index.tsx#L100 we can see that the TrainSessions are mapped to the bot's languages. Therefore, TrainSessions languages are the same as the bot's languages